src/platform/iotjs_module_adc-linux-general.inl.h

```diff
@@ -16,28 +16,95 @@
 #ifndef IOTJS_MODULE_ADC_LINUX_GENERAL_INL_H
 #define IOTJS_MODULE_ADC_LINUX_GENERAL_INL_H
 
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
 
+#include "iotjs_systemio-linux.h"
 #include "module/iotjs_module_adc.h"
 
 
+#define ADC_INTERFACE "/sys/devices/12d10000.adc/iio:device0/"
+#define ADC_PIN_INTERFACE "in_voltage%d_raw"
+#define ADC_PIN_FORMAT ADC_INTERFACE ADC_PIN_INTERFACE
+
+#define ADC_PATH_BUFFER_SIZE DEVICE_IO_PATH_BUFFER_SIZE
+#define ADC_PIN_BUFFER_SIZE DEVICE_IO_PIN_BUFFER_SIZE
+#define ADC_VALUE_BUFFER_SIZE 64
+
+
+// Implementation used here are based on:
+//  https://www.kernel.org/doc/Documentation/adc/sysfs.txt
+
+
+void InitializeadcWorker(uv_work_t* work_req) {
+  iotjs_adc_reqwrap_t* req_wrap = iotjs_adc_reqwrap_from_request(work_req);
+  iotjs_adc_reqdata_t* req_data = iotjs_adc_reqwrap_data(req_wrap);
+  DDDLOG("ADC InitializeadcWorker()");
+
+  // Check if ADC interface exits.
+  if (DeviceCheckPath(ADC_INTERFACE)) {
+    req_data->result = kAdcErrOk;
+  } else {
+    req_data->result = kAdcErrExport;
+  }
+}
+
+void ReadadcWorker(uv_work_t* work_req) {
+  iotjs_adc_reqwrap_t* req_wrap = iotjs_adc_reqwrap_from_request(work_req);
+  iotjs_adc_reqdata_t* req_data = iotjs_adc_reqwrap_data(req_wrap);
+  uint32_t pin = req_data->pin;
+  DDDLOG("adc ReadadcWorker() - pin: %d", pin);
+
+  char value_path[ADC_PATH_BUFFER_SIZE] = { 0 };
+  snprintf(value_path, sizeof(value_path), ADC_PIN_FORMAT, pin);
+
+  char buffer[ADC_VALUE_BUFFER_SIZE];
+  if (DeviceOpenReadClose(value_path, buffer, sizeof(buffer))) {
+    req_data->result = kAdcErrOk;
+    req_data->value = atoi(buffer);
+  } else {
+    req_data->result = kAdcErrRead;
+  }
+}
+
+
+
 void iotjs_adc_export_worker(uv_work_t* work_req) {
-  IOTJS_ASSERT(!"Not implemented");
+  InitializeadcWorker(work_req);
```

Comment:
Linux drivers for ADC, GPIO and PWM use sysfs:
https://en.wikipedia.org/wiki/Sysfs
Artik 
https://developer.artik.io/documentation/developer-guide/gpio/kernel-gpio.html

Kernel interface

GPIO:
http://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/tree/Documentation/gpio/sysfs.txt
PWM:
http://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/tree/Documentation/pwm.txt
http://developer.toradex.com/knowledge-base/pwm-(linux)
ADC:
http://developer.toradex.com/knowledge-base/adc-(linux)
https://github.com/torvalds/linux/tree/master/drivers/iio/adc


GPIO, PWM, ADC is 100% file system I/O - open/close, write/read files. Nuttix uses other approach but at least for Linux these modules should be pure JavaScript using:
https://github.com/Samsung/iotjs/blob/master/docs/api/IoT.js-API-File-System.md

This 100% JS solution will be portable for all Linux devices - PC, RPi2 and Artik 5xx, 7xx, 10xx board and also for OpenWRT. We don't care about CPU architecture in JS so it will be much easier to maintain.